gk/main.c

```diff
@@ -686,6 +845,495 @@ gk_conf_put(struct gk_config *gk_conf)
+		}
+
+		policy = add_prefix_policy(gw_addr, prefix_len, gk_conf);
+		if (policy == NULL) {
+			ret = -1;
+			goto out;
+		}
+
+		policy->action = action;
+		nexthop = &policy->u.nexthop;
+		nexthop->eth_cache.stale = true;
+
+		ret = convert_str_to_ip(gw_addr, &nexthop->ip_addr);
+		if (ret < 0)
+			goto out;
+
+		if (gw_type == AF_INET)
+			nexthop->eth_cache.eth_hdr.ether_type = ETHER_TYPE_IPv4;
+		else
+			nexthop->eth_cache.eth_hdr.ether_type = ETHER_TYPE_IPv6;
+
+		break;
+	}
+
+	case GK_FWD_GRANTOR: {
+		struct ipaddr gt_ip_addr;
+		struct ipaddr gw_ip_addr;
+		struct ip_flow *flow = NULL;
+
+ 		gt_type = get_ip_type(gt_addr);
+		if (gt_type == AF_INET)
+			prefix_len = sizeof(struct in_addr) * 8;
+		else if (gt_type == AF_INET6)
+			prefix_len = sizeof(struct in6_addr) * 8;
+		else {
+			ret = -1;
+			goto out;
+		}
+
+		policy = add_prefix_policy(gt_addr, prefix_len, gk_conf);
+		if (policy == NULL) {
+			ret = -1;
+			goto out;
+		}
+
+		ret = convert_str_to_ip(gt_addr, &gt_ip_addr);
+		if (ret < 0)
+			goto out;
+
+		policy->action = action;
+		flow = &policy->u.grantor.flow;
+
+		if (gt_type == AF_INET) {
+			flow->proto = ETHER_TYPE_IPv4;
+			flow->f.v4.src =
+				gk_conf->net->back.ip4_addr.s_addr;
+			flow->f.v4.dst = gt_ip_addr.ip.v4.s_addr;
+		} else {
+			flow->proto = ETHER_TYPE_IPv6;
+			rte_memcpy(flow->f.v6.src,
+				gk_conf->net->back.ip6_addr.s6_addr,
+				sizeof(flow->f.v6.src));
+			rte_memcpy(flow->f.v6.dst,
+				gt_ip_addr.ip.v6.s6_addr,
+				sizeof(flow->f.v6.dst));
+		}
+
+		/* Fill up the @next_fib field in the policy. */
+
+		/* The Grantor server is an neighbor. */
+		if (gw_addr == NULL) {
+			policy->u.grantor.eth_cache.stale = true;
+			policy->u.grantor.eth_cache.eth_hdr.ether_type =
+				gt_ip_addr.proto;
+			policy->u.grantor.next_fib =
+				gt_ip_addr.proto == ETHER_TYPE_IPv4 ?
+				ltbl->policy_tbl : ltbl->policy_tbl6;
+			policy->u.grantor.next_fib->ref_cnt++;
+		} else {
+			int next_fib_id = -1;
+			ret = convert_str_to_ip(gw_addr, &gw_ip_addr);
+			if (ret < 0)
+				goto out;
+
+			/* Find the policy entry for the gateway. */
+			if (gw_ip_addr.proto == ETHER_TYPE_IPv4) {
+				next_fib_id = lpm_lookup_ipv4(
+					ltbl->lpm,
+					gw_ip_addr.ip.v4.s_addr);
+				if (next_fib_id < 0) {
+					ret = -1;
+					goto out;
+				}
+
+				policy->u.grantor.next_fib =
+					&ltbl->policy_tbl[next_fib_id];
+				ltbl->policy_tbl[next_fib_id].ref_cnt++;
+			} else if (gw_ip_addr.proto == ETHER_TYPE_IPv6) {
+				next_fib_id = lpm_lookup_ipv6(
+					ltbl->lpm6,
+					gw_ip_addr.ip.v6.s6_addr);
+				if (next_fib_id < 0) {
+					ret = -1;
+					goto out;
+				}
+
+				policy->u.grantor.next_fib =
+					&ltbl->policy_tbl6[next_fib_id];
+				ltbl->policy_tbl6[next_fib_id].ref_cnt++;
+			} else {
+				ret = -1;
+				goto out;
+			}
+		}
+
+		break;
+	}
+
+	default:
+		return -1;
+	}
+
+	ret = 0;
+
+out:
+	return ret;
+}
+
+static int
+add_range_policy(struct lua_gk_policy *gk_policy, struct gk_config *gk_conf)
```

Comment:
If I understood it right, it's an IP prefix and this is one of the spots where we need to abandon the name "policy".